File: src/wiki/link_fix.py
"""Conservative auto-repair for broken internal wiki links."""

from __future__ import annotations

import difflib
import re
from pathlib import Path

from .audit import collect_broken_links
from .schemas import BrokenLink, BrokenLinkFix
from .config import WikiConfig
from .headings import GitHubHeadingSlugger
from .links import fragment_id, resolve_page_route, split_target
from .paths import iter_document_files, route_for_document_file

WIKILINK_FULL_REGEX = re.compile(r"\[\[([^\]|]+)(?:\|([^\]]*))?\]\]")
MARKDOWN_LINK_FULL_REGEX = re.compile(r"(!?\[[^\]]*\]\()([^)]+)(\))")
FUZZY_ROUTE_CUTOFF = 0.86


def find_broken_link_fixes(config: WikiConfig, file_filter: set[str] | None = None) -> list[BrokenLinkFix]:
    """Return only unambiguous broken-link repairs."""
    existing_routes = {
        route_for_document_file(config, file_path) for file_path in iter_document_files(config)
    }
    heading_ids_by_route = _heading_ids_by_route(config)
    fixes: list[BrokenLinkFix] = []

    for issue in collect_broken_links(config, file_filter=file_filter):
        if issue.issue_kind not in {"missing_document", "missing_heading"}:
            continue
        if issue.match_start is None or issue.match_end is None or issue.full_match is None:
            continue
        if issue.link_kind not in {"WikiLink", "Markdown link"}:
            continue

        replacement = _suggest_replacement(
            config,
            issue,
            existing_routes,
            heading_ids_by_route,
        )
        if replacement is not None:
            fixes.append(replacement)

    return fixes


def apply_broken_link_fixes(
    config: WikiConfig,
    fixes: list[BrokenLinkFix],
    *,
    dry_run: bool = False,
) -> list[Path]:
    """Apply broken-link fixes bottom-up within each file."""
    if not fixes:
        return []

    by_path: dict[Path, list[BrokenLinkFix]] = {}
    for fix in fixes:
        by_path.setdefault(fix.issue.source_path, []).append(fix)

    changed: list[Path] = []
    for file_path, file_fixes in by_path.items():
        content = file_path.read_text(encoding="utf-8")
        ordered = sorted(file_fixes, key=lambda item: item.issue.match_start or 0, reverse=True)
        for fix in ordered:
            start = fix.issue.match_start
            end = fix.issue.match_end
            if start is None or end is None or fix.issue.full_match is None:
                continue
            if content[start:end] != fix.issue.full_match:
                continue
            new_match = _replace_target_in_match(fix.issue, fix.replacement_target)
            content = content[:start] + new_match + content[end:]

        if dry_run:
            changed.append(file_path)
            continue
        file_path.write_text(content, encoding="utf-8")
        changed.append(file_path)

    return changed


def remaining_broken_links(
    config: WikiConfig,
    file_filter: set[str] | None = None,
    *,
    fixes: list[BrokenLinkFix] | None = None,
) -> list[BrokenLink]:
    """Return broken links that would still fail after applying fixes."""
    issues = collect_broken_links(config, file_filter=file_filter)
    if not fixes:
        return issues

    fixed_keys = {
        (fix.issue.source_path, fix.issue.match_start, fix.issue.match_end, fix.issue.raw_target)
        for fix in fixes
    }
    return [
        issue
        for issue in issues
        if (issue.source_path, issue.match_start, issue.match_end, issue.raw_target) not in fixed_keys
    ]


def _heading_ids_by_route(config: WikiConfig) -> dict[str, set[str]]:
    heading_ids: dict[str, set[str]] = {}
    slugger = GitHubHeadingSlugger()
    for file_path in iter_document_files(config):
        if file_path.suffix.lower() != ".md":
            continue
        route = route_for_document_file(config, file_path)
        body = file_path.read_text(encoding="utf-8")
        if body.startswith("---"):
            parts = body.split("---", 2)
            if len(parts) > 2:
                body = parts[2]
        ids: set[str] = set()
        for match in re.finditer(r"^(#{1,6})\s+(.+)$", body, flags=re.MULTILINE):
            ids.add(slugger.slug(match.group(2).strip()))
        heading_ids[route] = ids
    return heading_ids


def _suggest_replacement(
    config: WikiConfig,
    issue: BrokenLink,
    existing_routes: set[str],
    heading_ids_by_route: dict[str, set[str]],
) -> BrokenLinkFix | None:
    target = issue.raw_target
    page_part, fragment = split_target(target)

    if issue.issue_kind == "missing_document":
        replacement_page = _replacement_route(config, issue.source_route, page_part, existing_routes)
        if replacement_page is None:
            return None
        replacement_target = replacement_page
        if fragment:
            replacement_target = f"{replacement_page}#{fragment}"
        return BrokenLinkFix(
            issue=issue,
            replacement_target=replacement_target,
            description=f"{target} -> {replacement_target}",
        )

    route = resolve_page_route(issue.source_route, target)
    if route is None or route not in existing_routes or not fragment:
        return None
    replacement_fragment = _replacement_heading(fragment, heading_ids_by_route.get(route, set()))
    if replacement_fragment is None:
        return None
    replacement_target = f"{page_part}#{replacement_fragment}" if page_part else f"#{replacement_fragment}"
    return BrokenLinkFix(
        issue=issue,
        replacement_target=replacement_target,
        description=f"{target} -> {replacement_target}",
    )


def _replacement_route(
    config: WikiConfig,
    source_route: str,
    page_part: str,
    existing_routes: set[str],
) -> str | None:
    if not page_part:
        return None

    renamed = (config.link.renames or {}).get(page_part)
    if renamed and renamed in existing_routes:
        return renamed

    resolved = resolve_page_route(source_route, page_part)
    if resolved and resolved in existing_routes:
        return resolved

    normalized = page_part.replace(" ", "_")
    if normalized in existing_routes:
        return normalized

    matches = difflib.get_close_matches(page_part, sorted(existing_routes), n=2, cutoff=FUZZY_ROUTE_CUTOFF)
    if len(matches) == 1:
        return matches[0]
    return None


def _replacement_heading(fragment: str, heading_ids: set[str]) -> str | None:
    target_fragment = fragment_id(fragment)
    if target_fragment in heading_ids:
        return fragment
    matches = difflib.get_close_matches(target_fragment, sorted(heading_ids), n=2, cutoff=FUZZY_ROUTE_CUTOFF)
    if len(matches) == 1:
        return matches[0]
    return None


def _replace_target_in_match(issue: BrokenLink, replacement_target: str) -> str:
    full_match = issue.full_match or ""
    if issue.link_kind == "WikiLink":
        match = WIKILINK_FULL_REGEX.fullmatch(full_match)
        if not match:
            return full_match
        display = match.group(2)
        if display is not None:
            return f"[[{replacement_target}|{display}]]"
        return f"[[{replacement_target}]]"

    match = MARKDOWN_LINK_FULL_REGEX.fullmatch(full_match)
    if not match:
        return full_match
    return f"{match.group(1)}{replacement_target}{match.group(3)}"

File: src/wiki/link_fix.ts
import { extname } from "@std/path";
import { LinkIndex } from "./wiki_links.ts";
import type { Config } from "./config.ts";
import { splitFrontmatterText, WIKILINK_FULL_REGEX } from "./document.ts";

import { fragmentId, resolvePageRoute, splitTarget } from "./links.ts";
import { readTextTolerant } from "./parser.ts";
import { iterDocumentFiles, routeForDocumentFile } from "./paths.ts";
import { GitHubHeadingSlugger } from "./headings.ts";
import { getCloseMatches } from "./sequence_matcher.ts";
import type { BrokenLink, BrokenLinkFix } from "./schemas/domain.ts";

const MARKDOWN_LINK_FULL_RE = /^(!?\[[^\]]*\]\()([^)]+)(\))$/;
const FUZZY_ROUTE_CUTOFF = 0.86;

function compareCodePoints(left: string, right: string): number {
  const leftPoints = Array.from(left, (character) => character.codePointAt(0)!);
  const rightPoints = Array.from(
    right,
    (character) => character.codePointAt(0)!,
  );
  const sharedLength = Math.min(leftPoints.length, rightPoints.length);
  for (let index = 0; index < sharedLength; index += 1) {
    const difference = leftPoints[index]! - rightPoints[index]!;
    if (difference !== 0) return difference;
  }
  return leftPoints.length - rightPoints.length;
}

function headingIdsByRoute(config: Config): Map<string, Set<string>> {
  const byRoute = new Map<string, Set<string>>();
  for (const filePath of iterDocumentFiles(config)) {
    const route = routeForDocumentFile(config, filePath);
    const ids = new Set<string>();
    if (extname(filePath).toLowerCase() === ".md") {
      const body = splitFrontmatterText(readTextTolerant(filePath)).body;
      const slugger = new GitHubHeadingSlugger();
      for (const match of body.matchAll(/^(#{1,6})\s+(.+)$/gm)) {
        ids.add(slugger.slug((match[2] ?? "").trim()));
      }
    }
    byRoute.set(route, ids);
  }
  return byRoute;
}

function replacementRoute(
  config: Config,
  sourceRoute: string,
  pagePart: string,
  existingRoutes: ReadonlySet<string>,
): string | null {
  if (pagePart.length === 0) return null;
  const renamed = config.link.renames?.[pagePart];
  if (renamed !== undefined && existingRoutes.has(renamed)) return renamed;
  const resolved = resolvePageRoute(sourceRoute, pagePart);
  if (resolved !== null && existingRoutes.has(resolved)) return resolved;
  const normalized = pagePart.replaceAll(" ", "_");
  if (existingRoutes.has(normalized)) return normalized;
  const matches = getCloseMatches(
    pagePart,
    [...existingRoutes].sort(compareCodePoints),
    2,
    FUZZY_ROUTE_CUTOFF,
  );
  return matches.length === 1 ? matches[0]! : null;
}

function replacementHeading(
  fragment: string,
  headingIds: ReadonlySet<string>,
): string | null {
  const targetFragment = fragmentId(fragment);
  if (headingIds.has(targetFragment)) return fragment;
  const matches = getCloseMatches(
    targetFragment,
    [...headingIds].sort(compareCodePoints),
    2,
    FUZZY_ROUTE_CUTOFF,
  );
  return matches.length === 1 ? matches[0]! : null;
}

function suggestReplacement(
  config: Config,
  issue: BrokenLink,
  existingRoutes: ReadonlySet<string>,
  headingsByRoute: ReadonlyMap<string, ReadonlySet<string>>,
): BrokenLinkFix | null {
  const target = issue.raw_target;
  const [pagePart, fragment] = splitTarget(target);
  if (issue.issue_kind === "missing_document") {
    const replacementPage = replacementRoute(
      config,
      issue.source_route,
      pagePart,
      existingRoutes,
    );
    if (replacementPage === null) return null;
    const replacementTarget = fragment
      ? `${replacementPage}#${fragment}`
      : replacementPage;
    return {
      issue,
      replacement_target: replacementTarget,
      description: `${target} -> ${replacementTarget}`,
    };
  }

  const route = resolvePageRoute(issue.source_route, target);
  if (route === null || !existingRoutes.has(route) || !fragment) return null;
  const replacementFragment = replacementHeading(
    fragment,
    headingsByRoute.get(route) ?? new Set(),
  );
  if (replacementFragment === null) return null;
  const replacementTarget = pagePart
    ? `${pagePart}#${replacementFragment}`
    : `#${replacementFragment}`;
  return {
    issue,
    replacement_target: replacementTarget,
    description: `${target} -> ${replacementTarget}`,
  };
}

export function findBrokenLinkFixes(
  config: Config,
  fileFilter: ReadonlySet<string> | null = null,
): BrokenLinkFix[] {
  const existingRoutes = new Set(
    iterDocumentFiles(config).map((filePath) =>
      routeForDocumentFile(config, filePath)
    ),
  );
  const headingsByRoute = headingIdsByRoute(config);
  const issues = LinkIndex.fromConfig(config).brokenLinks(fileFilter);
  const fixes: BrokenLinkFix[] = [];
  for (const issue of issues) {
    if (
      issue.issue_kind !== "missing_document" &&
      issue.issue_kind !== "missing_heading"
    ) {
      continue;
    }
    if (
      issue.match_start === null || issue.match_end === null ||
      issue.full_match === null
    ) {
      continue;
    }
    if (issue.link_kind !== "WikiLink" && issue.link_kind !== "Markdown link") {
      continue;
    }
    const fix = suggestReplacement(
      config,
      issue,
      existingRoutes,
      headingsByRoute,
    );
    if (fix !== null) fixes.push(fix);
  }
  return fixes;
}

function replaceTargetInMatch(
  issue: BrokenLink,
  replacementTarget: string,
): string {
  const fullMatch = issue.full_match ?? "";
  if (issue.link_kind === "WikiLink") {
    const match = new RegExp(WIKILINK_FULL_REGEX.source).exec(fullMatch);
    if (match === null) return fullMatch;
    const display = match[2];
    return display === undefined
      ? `[[${replacementTarget}]]`
      : `[[${replacementTarget}|${display}]]`;
  }
  const match = MARKDOWN_LINK_FULL_RE.exec(fullMatch);
  if (match === null) return fullMatch;
  return `${match[1]}${replacementTarget}${match[3]}`;
}

export function applyBrokenLinkFixes(
  fixes: readonly BrokenLinkFix[],
  dryRun = false,
): string[] {
  const byPath = new Map<string, { path: string; fixes: BrokenLinkFix[] }>();
  for (const fix of fixes) {
    const key = fix.issue.source_path;
    const entry = byPath.get(key) ?? { path: fix.issue.source_path, fixes: [] };
    entry.fixes.push(fix);
    byPath.set(key, entry);
  }

  const changed: string[] = [];
  for (const { path, fixes: pathFixes } of byPath.values()) {
    let content = readTextTolerant(path);
    pathFixes.sort((left, right) =>
      (right.issue.match_start ?? 0) - (left.issue.match_start ?? 0)
    );
    for (const fix of pathFixes) {
      const { match_start: start, match_end: end, full_match: fullMatch } =
        fix.issue;
      if (start === null || end === null || fullMatch === null) continue;
      if (content.slice(start, end) !== fullMatch) continue;
      const replacement = replaceTargetInMatch(
        fix.issue,
        fix.replacement_target,
      );
      content = content.slice(0, start) + replacement + content.slice(end);
    }
    if (!dryRun) Deno.writeTextFileSync(path, content);
    changed.push(path);
  }
  return changed;
}

function issueKey(issue: BrokenLink): string {
  return [
    issue.source_path,
    issue.match_start,
    issue.match_end,
    issue.raw_target,
  ].join("\0");
}

export function remainingBrokenLinks(
  config: Config,
  fileFilter: ReadonlySet<string> | null = null,
  fixes: readonly BrokenLinkFix[] | null = null,
): BrokenLink[] {
  const issues = LinkIndex.fromConfig(config).brokenLinks(fileFilter);
  if (fixes === null || fixes.length === 0) return issues;
  const fixed = new Set(fixes.map((fix) => issueKey(fix.issue)));
  return issues.filter((issue) => !fixed.has(issueKey(issue)));
}
